File: 206.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        return nullptr;
    }
};

ListNode* change_node(ListNode* node) {
    cout << node << endl;
    node->val = 3;
    node = (node->next);
    node->val = 4;
    return node;
}

int main(int argc, char const *argv[])
{
    // ListNode* a = new ListNode(1);
    // ListNode* b = new ListNode(2);
    ListNode a(1), b(2);
    a.next = &b;
    ListNode c = a;
    
    cout << "a: " << &a << " b: " << &b << " c: " << &c << endl;
    auto d = change_node(&a);
    cout << "d: " << &d << endl;
    cout << a.val << b.val <<  d->val <<endl;
    // cout << a.val << b.val << change_node(a).val << a.val << endl;
    int num = 1;
    int* add = &num;
    cout << add << " " << *add << endl;
    cout << &add << endl;
    cout << &num;
    return 0;
}

File: 206.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # if not head or not head.next: return head
        # a, b = None, head
        # while b:
        #     tmp = b.next
        #     b.next = a
        #     a = b
        #     b = tmp
        # return a
        if not head or not head.next: return head
        a, b = None, head
        def reverse(a, b):
            if not b: return a
            tmp = b.next
            b.next = a
            # a, b = b, tmp
            return reverse(b, tmp)
        return reverse(a, b)


    def reverseList2(self, head: ListNode) -> ListNode:
        a, b = None, head
        while b:
            tmp = b.next
            b.next = a
            a = b
            b = tmp
        return a
            

def change_node(node):
    node = node.next
    return node

if __name__ == "__main__":
    a = ListNode(1)
    b = ListNode(2)
    a.next = b
    print(id(a), id(b), id(change_node(a)), id(a), a.val)

